### ISSUE TITLE
Bulk suppressions: a same-count swap in one file+rule bucket is invisible, and --prune-suppressions doesn't catch it

### ISSUE THREAD

**Issue by yunusdim** [eslint/eslint#21226]
### ESLint version

v10.8.1

### What problem do you want to solve?

The suppressions file (eslint-suppressions.json) records a count per (file, rule), not per specific violation. The docs already say this: suppressions are tracked per file and rule, not per line or code change. In practice that means one specific kind of change is invisible to both a normal lint run and to --prune-suppressions.

I reproduced it locally with ESLint 10.8.1. Two real no-unused-vars violations in one file (alpha, beta), suppressed with --suppress-all, giving count: 2 for that file+rule. Then I fixed alpha for real and introduced a different, unrelated no-unused-vars violation in the same file (gamma). beta was still there, unfixed. Total violations stayed at 2.

Running eslint with the suppressions file in place: clean, exit 0, nothing reported. Running eslint --prune-suppressions: no change to the file, no warning, nothing to prune, because the bucket is still exactly full at count 2, just covering a different combination of lines than before.

So a real fix and a real regression of the same rule, in the same file, cancel out completely, and there's no command in the tool, not even the one meant for auditing the suppressions file, that surfaces it.

### What do you think is the correct solution?

I don't think per-line tracking forever is the right ask, that's a much bigger design change and probably not worth it for what's meant to be a lightweight bulk-suppression tool. But --prune-suppressions already has to re-run the lint and compare against the suppressions file to decide what's stale. It could also flag, even just as a warning, buckets where the count still matches but the specific occurrences it's covering shifted (the file changed in a way that touches that rule, even though the recorded count didn't move). That wouldn't require storing per-line identity long-term, just a diff at prune time between the file's current state and its state when the suppression was last generated or pruned.

Short of that, it would help to say explicitly in the docs that --prune-suppressions won't catch this case, since the only caveat mentioned right now is about tracking granularity, not about what the audit command itself can and can't detect.

### Participation

- [ ] I am willing to submit a pull request for this change.

### AI acknowledgment

- [ ] I did not use AI to generate this issue report.
- [x] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

For context: I ran into this while looking at how a few different tools' baseline/ignore mechanisms handle the same underlying problem, a permitted-violations file that isn't tied to specific instances. Wrote it up more formally here if it's useful: https://doi.org/10.5281/zenodo.21908527

**Comment by fasttime** [comment]
Thanks for the issue @yunusdim. This is framed more as brainstorming than a change request, and there aren't any links or citations apart from the author's blog post, so it's a bit hard to turn it into an actionable task. In the absence of a concrete use case, we could perhaps update the documentation to clarify that ESLint tracks the number of suppressions per file and rule, without diving into the caveats of `--prune-suppressions` or other details of the current implementation. I don't think we can make any broader changes without a problem to solve.

**Comment by yunusdim** [comment]
Fair on the framing — I opened this as a problem statement rather than a change request, and that's on me. Here is the concrete case and a specific proposal.

**Repro, on v10.9.1:**

```
src/app.js:  const alpha = 1;  const beta = 2;
eslint src/app.js --suppress-all        → {"src/app.js":{"no-unused-vars":{"count":2}}}

# fix alpha for real; introduce an unrelated gamma; beta untouched
src/app.js:  const beta = 2;  const gamma = 3;

eslint src/app.js                       → clean, exit 0
eslint src/app.js --prune-suppressions  → no change, exit 0, nothing pruned
```

Two controls, same setup, showing the mechanism is otherwise working — which is why I think this is a gap rather than intended behaviour:

```
count increases 1→3          → all three reported, exit 1
count decreases 2→1          → --prune-suppressions rewrites to 1
count unchanged, set swapped → nothing, in either command
```

**The use case.** A team adopts bulk suppressions on a legacy file to unblock CI, intending to burn the bucket down. Someone fixes one violation and, in the same PR or a later one, writes a new one of the same rule in the same file. CI is green, the count is unchanged, and `--prune-suppressions` reports nothing to prune — so the usual signal that the debt is shrinking never fires, and the bucket can be recycled indefinitely without anyone choosing that. The failure is silent in both directions: the team gets no credit for the fix and no warning about the regression.

**Concrete request, minimal and backwards-compatible.** Keep `count` as the gate, unchanged, so no existing run changes behaviour. Alongside it, record a content-derived digest for each `(file, rule)` bucket — for example a hash over the sorted violation messages — and have `--prune-suppressions` report when the digest changes while the count does not. Not line-based, so edits elsewhere in the file still don't break matching, which I take to be the reason the current design is per file and rule.

If a digest is more than you want to carry, the smaller version is a flag that makes `--prune-suppressions` report bucket contents that changed identity, without failing the run.

Documentation would help, but on its own it tells people the signal is unreliable without giving them a way to see when it has gone unreliable.

**Comment by fasttime** [comment]
I think that violation messages are generally not granular enough to distinguish unrelated reports. Many rules such as `callback-function` simply report the same message(s) regardless of the content. Another problem is that violation messages may change across different versions of a rule, and when that happens, users expect not to lose their suppressions from previous runs, even when using `--prune-suppressions`. How do other tools handle obsolete suppressions?

**Comment by yunusdim** [comment]
@fasttime both objections are right, and the first one sinks the shape I proposed. Violation messages are not identity — `no-unused-vars` says the same thing about every variable it finds, so hashing that gives you a digest of the rule, not of what is suppressed. Drop that part.

On your question, since it is the useful one. Five tools in this space, and none of them uses the message.

**jscpd** shipped exactly this in `v5.1.0` two weeks ago. The baseline stores a fingerprint per clone pair built from a content hash of the duplicated fragment plus the path — not line numbers, not descriptions. `--fail-on-new-clones` gates on a fingerprint being absent from the baseline rather than on the aggregate percentage, and `--update-baseline` prints added and removed fingerprint counts, so the identity change stays visible in CI logs even when it is accepted.

**SwiftLint**'s baseline anchors on the rule reason plus the text of the offending line, and falls back to a displacement-tolerant path when no anchor survives — which is where a same-run swap becomes invisible again.

**gitleaks** fingerprints as `file:rule:index`, where the index is the occurrence number. Replace a secret with a different one on the same line and the fingerprint is unchanged, so the suppression still covers it.

**bandit**'s baseline compares issue equality with no multiplicity, so a one-entry baseline absorbs two new violations while the metrics footer still counts them.

**dependency-cruiser**'s maintainer put the problem precisely while scoping a shrink-only mode: the comparison function is the feature, because it defines what counts as the same violation between runs — and a rule whose severity changed is still the same rule for that purpose, but not for the existing comparator.

So the field converges on one answer: hash the content that was suppressed, never the description of it. That also answers your second objection, and it is why content is the right axis rather than a workaround. A digest over the normalised source text in the reported range does not move when a rule's message is reworded, or when the rule is upgraded — which is the case where you are right that users must not lose suppressions. It moves when the suppressed code moves, which is the event that is currently silent.

Concretely, and still backwards compatible: keep `count` as the gate, unchanged. Alongside it, store a digest over the normalised text of the suppressed ranges for each `(file, rule)` bucket, and have `--prune-suppressions` report when the digest changed while the count did not. It reports, it does not invalidate — no existing run changes behaviour and no suppression is lost.

For where the comparison comes from: I have been reporting this same failure across assurance tooling, forty-three cases, public, including the ones that were rejected and why. The ledger is at https://github.com/yunusdim/casos-defectos-estructurales, and the formal treatment of when a criterion like this can be trivialised is at https://doi.org/10.5281/zenodo.21908527.

**Comment by fasttime** [comment]
Thanks. I understand that your suggestion is to use a hash of the _content_, that is, the normalized source text in the reported range, to identify each suppression more reliably. Aside from determining how the source text should be normalized, implementing this change would require an RFC. If you are interested in seeing this change discussed by the team, I suggest updating the issue description with your concrete proposal, as that would help focus the discussion. For reference, you may also want to read #16755, which was filed by another blog author and provides an overview of the approaches that existed at the time. The discussion eventually led to the current implementation of bulk suppressions in ESLint.
